Clip against near plane for bounds calculus

kvark opened the pull request:
Fixes #2389
Fixes #2272
Improves #2429 a great deal (there are general plane splitting issues in there still)

TODO:
- [x] reviews
- [x] green CI
- [x] try push: https://treeherder.mozilla.org/#/jobs?repo=try&revision=ddccbd9f16aea173fe950293f1a4f429650dea6a

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/2741)
<!-- Reviewable:end -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #2736) made this pull request unmergeable. Please resolve the merge conflicts.

kvark commented:
Inviting reviewers! cc @glennw

mrobinson commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -369,13 +369,20 @@ impl ClipScrollNode {
         // use the BorderCorner clip type and they always have at last one non-ClipOut
         // Rectangle ClipSource.
-        let screen_outer_rect = screen_outer_rect.expect("Clipping node didn't have outer rect.");
+        let screen_outer_rect = match screen_outer_rect {
```
Do ClipScrollNodes not always have an outer rectangle now?

kvark replied:
In this code, it may be `None` if either `local_outer_rect` of the clip source is `None`, or the computed bounding rect is outside the screen.

mrobinson commented on webrender/src/util.rs:
```diff
@@ -160,21 +162,114 @@ pub fn calculate_screen_bounding_rect(
-        transform.transform_point2d(&rect.bottom_left()),
-        transform.transform_point2d(&rect.bottom_right()),
+    screen_bounds: Option<&DeviceIntRect>,
+) -> Option<DeviceIntRect> {
```
I recall this function as being extremely performance sensitive, so is there any way to have a fast path when it isn't necessary to do all of these calculations?

kvark replied:
The way it's build is to avoid extra steps unless necessary. So the old `transform_point2d` for each corner is equivalent to `transform_point_2d_homogeneous` followed by `to_point2d` at the end.
Extra overhead may come from the following sources:
  1. more stuff being moved around (copies are the doom of Rust code)
  2. comparing 4 f32 values to 0 in `homogens.iter().any(|h| h.w <= 0.0)`
  3. worse instruction cache utilization for the function body being grown

Overall, I think it's a reasonable compromise, and it's not clear to me if we can do better here.

kvark commented:
I'm a bit puzzled over AppVeyor reftest failure...

kvark commented:
@mrobinson thanks for taking a look! I think we are all set, unless there are more concerns.

mrobinson commented on wrench/src/main.rs:
```diff
@@ -15,7 +15,7 @@ extern crate core_graphics;
 #[cfg(target_os = "windows")]
 extern crate dwrote;
-#[cfg(feature = "logging")]
+#[cfg(feature = "env_logger")]
```
I think this was landed in another patch?

kvark commented:
@mrobinson 
Rebased and fixed the local outer rect now (hopefully).

> One thing we could do here is to detect when transformation doesn't project into three dimensions and to use the old code for this situation. You could have this path be an early return and avoid the long if statement in the body.

It would involve more logic duplication and control flow tweaks, so I'm not sure it's worth it. Maybe we can leave a TODO comment in the code saying to keep an eye for this branch when profiling?

kvark commented:
Updated try push: https://treeherder.mozilla.org/#/jobs?repo=try&revision=7dd1dfe5c8453fe34e37bcfbfd23f54e4bcc9a2e

kvark commented:
I don't want this to stall any longer.
@bors-servo r=mrobinson

bors-servo commented:
:pushpin: Commit cafeb7a has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson cafeb7a0644cdbcbfb574b08a52057a6eb3b4ebd -->

bors-servo commented:
:hourglass: Testing commit cafeb7a0644cdbcbfb574b08a52057a6eb3b4ebd with merge fac0040b4597d6c4a4b36fffbb8117782027e7ca...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.9301), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/YUebX3ULTZC9CzSRlvPnmw)
Approved by: mrobinson
Pushing fac0040b4597d6c4a4b36fffbb8117782027e7ca to master...
